### ISSUE TITLE
Extend CVC4 support for Non Linear Real Arithmetic

### ISSUE THREAD

**Issue by ghost** [pysmt/pysmt#318]
I get the following error when I pass a QF_NRA benchmark to cvc4 solver.
I think CVC4 can handle QF_NRA benchmarks. 
So reporting this bug.

```
Traceback (most recent call last):
  File "/es0/irfan/pysmt_nra_smt/bin/pysmt_cvc4.py", line 18, in <module>
    main()
  File "/es0/irfan/pysmt_nra_smt/bin/pysmt_cvc4.py", line 11, in main
    cvc4.add_assertion(f)
  File "/es0/irfan/local/pysmt35/pysmt/solvers/cvc4.py", line 81, in add_assertion
    term = self.converter.convert(formula)
  File "/es0/irfan/local/pysmt35/pysmt/decorators.py", line 83, in catch_conversion_error_wrap
    res = f(*args, **kwargs)
  File "/es0/irfan/local/pysmt35/pysmt/solvers/cvc4.py", line 214, in convert
    return self.walk(formula)
  File "/es0/irfan/local/pysmt35/pysmt/walkers/dag.py", line 111, in walk
    res = self.iter_walk(formula, **kwargs)
  File "/es0/irfan/local/pysmt35/pysmt/walkers/dag.py", line 103, in iter_walk
    self._process_stack(**kwargs)
  File "/es0/irfan/local/pysmt35/pysmt/walkers/dag.py", line 96, in _process_stack
    self._compute_node_result(formula, **kwargs)
  File "/es0/irfan/local/pysmt35/pysmt/walkers/dag.py", line 80, in _compute_node_result
    self.memoization[key] = f(formula, args=args, **kwargs)
  File "/es0/irfan/local/pysmt35/pysmt/solvers/cvc4.py", line 294, in walk_times
    raise NonLinearError(formula)
pysmt.exceptions.NonLinearError: (b.speed_x__AT7 * (-1.0 * b.speed_x__AT7))
```

**Comment by marcogario** [comment]
The message that you are getting is by design. If you were to instantiate the solver by specifying that you will be using QF_NRA, you would get an exception telling you that the logic is not supported by the solver. 
In order to provide a reasonable interface across solvers, we need to limit the scope of which theories are supported by whom. Which logics are currently supported is described in [1] (or you can look at Solver.LOGICS within the source code of each solver).

Now the question is whether we should extend the CVC4 interface to support NRA. According to the documentation [2] (that might be out-dated), support for NRA is very limited. By looking at the results of the SMT-COMP of this year [3] we can see that the techniques implemented there managed to solve ~25% of the instances (out of 10k instances). To put this into perspective, Yices2 and Z3 solved ~98% of the instances.

In conclusion, this is not going to be addressed in the near future. The next QF_NRA solver is going to be Yices2. 

If you really need to access CVC4, then use the SMT-LIB interface. Otherwise, if you are interested in extending pySMT to allow CVC4to work on QF_NRA, let me know, I'll guide you through the steps.


[1] https://github.com/pysmt/pysmt/blob/master/README.rst
[2] http://cvc4.cs.nyu.edu/wiki/User_Manual#Supported_Theories
[3] http://smtcomp.sourceforge.net/2016/results-QF_NRA.shtml?v=1467876482

**Comment by ahmedirfan1983** [comment]
I agree with you. 

I can try to run cvc4-nra outside pysmt. 

You can close the issue if you want.
